repos/system_upgrade/el7toel8/actors/cupscheck/actor.py

```diff
@@ -0,0 +1,226 @@
+"""
+CUPSD_CONF = '/etc/cups/cupsd.conf'
+CUPSFILES_CONF = '/etc/cups/cups-files.conf'
+
+"""
+Names for better navigation through change list
+"""
+KEY = 0
+DIRECTIVES = 1
+SOURCE = 2
+TARGET = 3
+
+"""
+List of changed directives
+
+Scheme:
+Dictionary key # Directives # Input configuration file # Output configuration file
+
+Some changes happened only in one configuration file, so the last member of list is 'None'.
+"""
+cupsd_changes_list = [
+                      ['include', ['Include'], CUPSD_CONF, None],
+                      ['digest', ['Digest', 'BasicDigest'], CUPSD_CONF, None],
+                      ['env', ['SetEnv', 'PassEnv'], CUPSD_CONF, CUPSFILES_CONF],
+                      ['printcap', ['PrintcapFormat'], CUPSD_CONF, CUPSFILES_CONF]
+                    ]
+
+cupsfiles_changes_list = [
+                          ['certkey', ['ServerCertificate', 'ServerKey'], CUPSFILES_CONF, None]
+                          ]
+
+class CupsCheck(Actor):
+    """
+    cups_check actor
+
+    Actor checks if cups package is installed and if one or more following
+    situations appears in configuration files:
+    - interface scripts
+    - use of 'Digest' or 'BasicDigest' authentication
+    - use of 'Include' directive
+    - use of 'ServerCertificate' and 'ServerKey' directives
+    - use of 'SetEnv' or 'PassEnv' directives
+    - use of 'PrintcapFormat' directive
+
+    The actor creates python dictionary from gathered data.
+    """
+
+    name = 'cups_check'
+    consumes = (InstalledRedHatSignedRPM,)
+    produces = (Report, Cupsmigrationmodel)
+    tags = (ChecksPhaseTag, IPUWorkflowTag)
+
+    def has_interface_script(self):
+      """
+      has_interface_script
+
+      Checks if any file is in /etc/cups/interfaces, which means there could be
+      print queues using interface script.
+
+      returns boolean
+      """
+      if os.path.exists('/etc/cups/interfaces'):
+        if len(os.listdir('/etc/cups/interfaces')) > 0:
+          return True
+
+      return False
+
+    def has_directive(self, directives_list, filename):
+      """
+      has_directive
+
+      Checks if configuration file has directive set.
+
+      parameters:
+      - directives_list - list of strings
+      - filename - string
+
+      returns boolean
+      """
+      found = False
+
+      with open(filename, 'r') as opened_config:
+        file_contents = opened_config.read()
+        for directive in directives_list:
+          # Digest/BasicDigest are directive values, so regular expression differs
+          if directive == 'Digest' or directive == 'BasicDigest':
+            regexp_string = '\s+'
+          else:
+            regexp_string = '^\s*'
+          regexp_string = regexp_string + directive + '\s'
+          pattern = compile(regexp_string, MULTILINE)
+          matched_list = findall(pattern, file_contents)
+          if len(matched_list) > 0:
+            found = True
+            break
+        opened_config.close()
```

Comment:
It is not necessary to close() opened_config. After the code exits current `with` context, the opened file will be closed automatically.

Reply:
Noted. Fixed.